### ISSUE TITLE
Using intervals to track types causes verification failures

### ISSUE THREAD

**Issue by dthaler** [vbpf/ebpf-verifier#286]
Symptom:  "
Where code is:
```
2171:
  assert r7.type in {ctx, stack, packet, shared};
  assert valid_access(r7.offset+32, width=4);
  r2 = *(u32 *)(r7 + 32);
  goto 2172;
```

Analysis:
Above happens after a join between two paths:
Path A: r7.type=shared, r7.offset=0, packet_size=[34, 65534]
Path B: r7.type=stack, r7.offset=392, packet_size=[34, 65534]
After the join, this combines to:
              r7.type=[-2, 0], r7.offset=[0, 392], packet_size=[34, 65534]

Now the problem is the type range where [-2,0] means { stack, packet, shared }.
Notice that only stack and shared are possible, not packet.
Since the joined state has packet in the list of possibilities, it checks the offset against
packet size and determines it is not safe.   This is an artifact of the fact that the type
is a range, and the values are ordered such that packet is in the middle.

It seems like this could be a common problem with the current verifier code.

The best option would be for the verifier to use a set, rather than a range interval, to track types.
This is similar to the discussion PR #277 about map fds,
but the case here is even stronger.   Any time the region size of an unused type in the middle
of an integer has the smallest region size, this issue will be hit.

**Comment by elazarg** [comment]
No. It should be an equality-relational domain. I've started implementing it a couple of months ago and it got stale. If it's important I can get back to it.
